apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java

```diff
@@ -0,0 +1,144 @@
+        // shard archive data has been downloaded at that point and stored (unpacked?) in configured folder
+        String zipFileName = shardPresentData.getFileIdValue();
+        Path zipInFolder = csvImporter.getDataExportPath().resolve(zipFileName + ".zip").toAbsolutePath();
+        log.debug("Try unpack file name '{}'", zipInFolder);
+        boolean unpackResult = zipComponent.extract(zipFileName + ".zip", csvImporter.getDataExportPath().toString());
+        log.debug("Zip is unpacked = {}", unpackResult);
+        Genesis.apply(); // import public data - genesis public Keys + balances
+
+        // import additional tables
+        List<String> tables = List.of(ShardConstants.SHARD_TABLE_NAME,
+                ShardConstants.BLOCK_TABLE_NAME, ShardConstants.TRANSACTION_TABLE_NAME,
+                ShardConstants.TRANSACTION_INDEX_TABLE_NAME, ShardConstants.BLOCK_INDEX_TABLE_NAME);
+        for (String table : tables) {
+            try {
+                long rowsImported = csvImporter.importCsv(table, ShardConstants.DEFAULT_COMMIT_BATCH_SIZE, true);
+                log.debug("Imported '{}' rows = {}", ShardConstants.SHARD_TABLE_NAME, rowsImported);
+            } catch (Exception e) {
+                log.error("CSV import error for '{}', RETURN.......", table, e);
+                return;
+            }
+        }
+        // import derived tables
+        Collection<String> tableNames = derivedTablesRegistry.getDerivedTables().stream().map(Object::toString).collect(Collectors.toList());
+        for (String table : tableNames) {
+            try {
+                long rowsImported = csvImporter.importCsv(table, 100, true);
+                log.debug("Imported '{}' rows = {}", table, rowsImported);
+            } catch (Exception e) {
+                log.error("CSV import error for '{}', RETURN.......", table, e);
+                return;
+            }
+        }
+        // set to start work block download thread (starting from shard's snapshot block here)
+        blockchainProcessor.setGetMoreBlocks(true); // IMPORTANT CALL !!!
+    }
+
+    /**
+     * Process event when no sharding is present in network and we should go 'old scenario'
+     *
+     * @param shardPresentData not used actually
+     */
+    public void onNoShardPresent(@ObservesAsync @ShardPresentEvent(ShardPresentEventType.NO_SHARD) ShardPresentData shardPresentData) {
+        // start adding old Genesis Data
+        log.info("Genesis block not in database, starting from scratch");
+        TransactionalDataSource dataSource = databaseManager.getDataSource();
+        Connection con = dataSource.begin();
+        try {
+            Block genesisBlock = Genesis.newGenesisBlock();
+            addBlock(dataSource, genesisBlock);
+            long initialBlockId = genesisBlock.getId();
+            log.debug("Generated Genesis block with Id = {}", initialBlockId);
```

Comment:
Looks like this code is copypasted from BlockchainProcessorImpl.addGenesisBlock

Reply:
**addGenesisBlock()**  was renamed to: 
**private void continuedDownloadOrTryImportGenesisShard()**
Those functionality was moved from BlockchainProcessorImpl to that class.